cmd/kubeadm/app/cmd/config.go

```diff
@@ -272,6 +273,48 @@ func newCmdConfigMigrate(out io.Writer) *cobra.Command {
 	return cmd
 }
 
+// newCmdConfigValidate returns cobra.Command for the "kubeadm config validate" command
+func newCmdConfigValidate() *cobra.Command {
+	var cfgPath string
+
+	cmd := &cobra.Command{
+		Use:   "validate",
+		Short: "Read a file containing the kubeadm configuration API and report any validation problems",
+		Long: fmt.Sprintf(dedent.Dedent(`
+			This command lets you validate a kubeadm configuration API file and report any warnings and errors.
+			If there are no errors the exit status will be zero, otherwise it will be non-zero.
+			Any unmarshaling problems such as unknown API fields will trigger warnings. Unknown API versions and
+			fields with invalid values will trigger fatal errors. Any other errors or warnings may be reported
+			depending on contents of the input file.
+
+			In this version of kubeadm, the following API versions are supported:
+			- %s
+		`), kubeadmapiv1.SchemeGroupVersion),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(cfgPath) == 0 {
+				return errors.Errorf("the --%s flag is mandatory", options.CfgPath)
+			}
+
+			cfgBytes, err := os.ReadFile(cfgPath)
+			if err != nil {
+				return err
+			}
+
+			// MigrateOldConfig performs conversion to internal type, strict unmarshaling,
+			// defaulting and validation. It can be used for the validation purpose with output bytes
+			// ignored.
+			if _, err = configutil.MigrateOldConfig(cfgBytes); err != nil {
```

Comment:
this is essentially reusing the migrate backend and ignoring the output.
LMK, if you think we should refactor it for more clarity.

Reply:
I tried this command, one thing might worth to discuss here is that if only unknown obj defined in the cfg, there is no warning at all.

e.g. the config is defined as this, 

```
kind: unknownKind
apiVersion: kubeadm.k8s.io/v1beta3
unknowkey: unknownvalue
```
```
kubeadm config validate --config /home/davche02/backup/config/kubeadm/bootstrapfake.cfg
```
validation passed in this case.
We'd better to thrown a warning here.

Reply:
we should throw an error for unknown GVK.
IIRC, the code paths for init, join do that.

Reply:
> +1 to define another func for this command even if it just calls `MigrateOldConfig` instead.

there could be migrate and validate funcs that share a common backend.

the validate func will not have the marshaling / output.

Reply:
> we should throw an error for unknown GVK. IIRC, the code paths for init, join do that.

yes, the problem only exists if the cfg just contains a unknown GVK, let's see is there any chance to improve it.

Reply:
@neolit123 I just found there is a comments about this, 
https://github.com/kubernetes/kubernetes/blob/e3db9232d9fffbb809c5e0652f56a9eb819aef98/cmd/kubeadm/app/util/config/common.go#L194

Looks like unknown kinds are just ignored as intended, although I don't know why there is no error nor warning there. 
But if we are going to reuse that method, we should update the method to thrown an error for validation I think.

Reply:
yes, agreed that we should throw errors.
i will include such a  change as a separate commit here.

Reply:
> Looks like unknown kinds are just ignored as intended, although I don't know why there is no error nor warning there.

@chendave if you feed something like:

```
---
apiVersion: foo/v1beta1
kind: Bar
```

it throws
```
W0522 12:52:55.224931   15193 initconfiguration.go:306] unknown configuration schema.GroupVersionKind{Group:"foo", Version:"v1beta1", Kind:"Bar"}
W0522 12:52:55.225030   15193 initconfiguration.go:332] [config] WARNING: Ignored YAML document with GroupVersionKind foo/v1beta1, Kind=Bar
```

so warnings are already in place.
we can keep them as warnings unless users want errors..

Reply:
>  // Only kubeadm kinds are migrated. Others are silently ignored. 

updated this related comment to reduce confusion.

Reply:
> +1 to define another func for this command even if it just calls MigrateOldConfig instead.

added a new ValidateConfig util function now.

Reply:
> ```
> ---
> apiVersion: foo/v1beta1
> kind: Bar
> ```

Should we add this to the unit test case?

Reply:
With the latest code https://github.com/kubernetes/kubernetes/pull/118013/commits/0d85d67d20da4174c4db71e2565f94ad9661abda, there is no warning if the configuration is not match Join/Init/Cluster according to https://github.com/kubernetes/kubernetes/pull/118013/files#diff-5569e71dfa98866b583171d92e018cf11680c6e1ab16e55ecd0c1e2a0f1f4f4eR250-R264.

Reply:
> Should we add this to the unit test case?

it's just a warning on stderr (klog) so there is no straightforward way to test it. we could switch it to error and it would be easier to unit test.

> With the latest code https://github.com/kubernetes/kubernetes/commit/0d85d67d20da4174c4db71e2565f94ad9661abda, there is no warning

i think the problem is that i did not test just unknown GVK. it was combined with known GVK. the logic needs update..

Reply:
updated, see new PR description and release note amend.
- unknown APIs (GVK) and unknown fields now throw errors in migrate/validate commands
- other commands like init/join are not affected
- added better unit tests for both new commands
-